java-checks/src/main/java/org/sonar/java/checks/security/PubliclyWritableDirectoriesCheck.java

```diff
@@ -0,0 +1,169 @@
+    .names("createTempDirectory")
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers NIO_TEMP_FILE_MATCHER = MethodMatchers.create()
+    .ofTypes("java.nio.file.Files")
+    .names("createTempFile")
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers MAP_GET = MethodMatchers.create()
+    .ofSubTypes("java.util.Map")
+    .names("get")
+    .addParametersMatcher("java.lang.Object")
+    .build();
+
+  private static final MethodMatchers SYSTEM_GETENV = MethodMatchers.create()
+    .ofSubTypes("java.lang.System")
+    .names("getenv")
+    .addWithoutParametersMatcher()
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree mit = (MethodInvocationTree) tree;
+      if (createdInTempDir(mit) || hasSensitiveFileName(mit) || usesSystemTempDir(mit)) {
+        reportIssue(tree, MESSAGE);
+      }
+    } else {
+      NewClassTree newClassTree = (NewClassTree) tree;
+      if (CREATE_FILE_MATCHERS.matches(newClassTree) &&
+        isSensitiveFileName(newClassTree.arguments().get(0))) {
+        reportIssue(tree, MESSAGE);
+      }
+    }
+  }
+
+  private static boolean hasSensitiveFileName(MethodInvocationTree mit) {
+    return CREATE_FILE_MATCHERS.matches(mit) &&
+      isSensitiveFileName(mit.arguments().get(0));
+  }
+
+  private static boolean usesSystemTempDir(MethodInvocationTree mit) {
+    return MAP_GET.matches(mit) && hasTMPAsArgument(mit) && isInitializedWithSystemGetEnv(mit);
+  }
+
+  private static boolean hasTMPAsArgument(MethodInvocationTree mit) {
+    return mit.arguments().get(0).asConstant(String.class)
+      .map(TMP_DIR_ENV::contains)
+      .orElse(false);
+  }
+
+  private static boolean createdInTempDir(MethodInvocationTree mit) {
+    return TEMP_DIR_MATCHER.matches(mit) ||
+      (NIO_TEMP_DIR_MATCHER.matches(mit) && (mit.arguments().size() == 1)) ||
+      (NIO_TEMP_FILE_MATCHER.matches(mit) && (mit.arguments().size() == 2));
+  }
+
+  private static boolean isSensitiveFileName(ExpressionTree expressionTree) {
+    return expressionTree.asConstant(String.class)
+      .filter(path -> PUBLIC_WRITABLE_DIRS.stream().anyMatch(path::startsWith))
+      .isPresent();
+  }
+
+  private static boolean isInitializedWithSystemGetEnv(MethodInvocationTree mit) {
+    return getInvokedSymbol(mit)
+      .filter(ExpressionsHelper::isNotReassigned)
+      .map(Symbol::declaration)
+      .filter(decl -> decl.is(Tree.Kind.VARIABLE))
+      .map(VariableTree.class::cast)
+      .map(VariableTree::initializer)
+      .map(ExpressionUtils::skipParentheses)
+      .filter(initializer -> initializer.is(Tree.Kind.METHOD_INVOCATION))
+      .map(MethodInvocationTree.class::cast)
+      .map(SYSTEM_GETENV::matches)
+      .orElse(false);
```

Comment:
😄

Reply:
I've tried to write it firstly using imperative way, and the code looked very complicated and had a lot of null checks. So I decided to apply declarative approach here.

As for me, it looks quite readable and understandable, but feel free to challenge this approach. Cause it might be just my Scala brain deformation 😅

Reply:
I also like this code... I probably have the same brain deformation though.